robot_calibration/include/robot_calibration/ceres/chain3d_to_mesh_error.h

```diff
@@ -0,0 +1,162 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef ROBOT_CALIBRATION_CERES_CHAIN3D_TO_MESH_ERROR_H
+#define ROBOT_CALIBRATION_CERES_CHAIN3D_TO_MESH_ERROR_H
+
+#include <limits>
+#include <string>
+#include <math.h>
+#include <ceres/ceres.h>
+#include <robot_calibration/calibration_offset_parser.h>
+#include <robot_calibration/mesh_loader.h>
+#include <robot_calibration/ceres/calibration_data_helpers.h>
+#include <robot_calibration/models/camera3d.h>
+#include <robot_calibration/models/chain.h>
+#include <robot_calibration_msgs/CalibrationData.h>
+
+namespace robot_calibration
+{
+
+/**
+ * @brief Get the squared distance line segment A-B for point C
+ *
+ * Based on "Real Time Collision Detection", pg 130
+ */
+double distToLine(Eigen::Vector3d& a, Eigen::Vector3d& b, Eigen::Vector3d c)
+{
+  Eigen::Vector3d ab = b - a;
+  Eigen::Vector3d ac = c - a;
+  Eigen::Vector3d bc = c - b;
+
+  double e = ac.dot(ab);
+  if (e <= 0.0)
+  {
+    // Point A is closest to C
+    return ac.dot(ac);
+  }
+  double f = ab.dot(ab);
+  if (e >= f)
+  {
+    // Point B is closest to C
+    return bc.dot(bc);
+  }
+  // C actually projects between
+  return ac.dot(ac) - e * e / f;
+}
+
+/**
+ *  \brief Error block for computing the fit between a set of projected
+ *         points and a mesh (usually part of the robot body). Typically used
+ *         to align sensor with the robot footprint.
+ */
+struct Chain3dToMesh
+{
+  /**
+   *  \brief This function is not used direcly, instead use the Create() function.
+   *  \param chain_model The model for the chain, used for reprojection.
+   *  \param offsets Easy access to the free parameters.
+   *  \param data The calibration data collected.
+   *  \param mesh_path Path to the mesh file to test against
+   */
+  Chain3dToMesh(ChainModel* chain_model,
+                CalibrationOffsetParser* offsets,
+                robot_calibration_msgs::CalibrationData& data,
+                MeshPtr& mesh)
+  {
+    chain_model_ = chain_model;
+    offsets_ = offsets;
+    data_ = data;
+    mesh_ = mesh;
+  }
+
+  virtual ~Chain3dToMesh() {}
+
+  bool operator()(double const * const * free_params,
+                  double* residuals) const
+  {
+    // Update calibration offsets based on free params
+    offsets_->update(free_params[0]);
+
+    // Project the camera observations
+    std::vector<geometry_msgs::PointStamped> chain_pts =
+        chain_model_->project(data_, *offsets_);
+
+    // Compute residuals
+    for (size_t i = 0; i < chain_pts.size() ; ++i)
+    {
+      Eigen::Vector3d p(chain_pts[i].point.x, chain_pts[i].point.y, chain_pts[i].point.z);
+
+      // Find shortest distance to any line segment forming a triangle
+      double dist = std::numeric_limits<double>::max();
+      for (size_t t = 0; t < mesh_->triangle_count; ++t)
```

Comment:
a bit picky but it would be great if you could replace 'i" and "t" with chain_pt_index and mesh_triangle_index for readability